src/tracer/MARBL_tracers.F90

```diff
@@ -0,0 +1,734 @@
+  character(len=48) :: flux_units ! The units for age tracer fluxes, either
+                                ! years m3 s-1 or years kg s-1.
+  logical :: OK
+  integer :: i, j, k, m, diag_size
+
+  if (.not.associated(CS)) return
+  if (CS%ntr < 1) return
+
+  CS%diag => diag
+
+  ! Register diagnostics (surface flux first, then interior tendency)
+  call register_MARBL_diags(marbl_instances%surface_flux_diags, diag, day, G, CS%surface_flux_diags)
+  ! call register_MARBL_diags(marbl_instances%interior_tendency_diags, diag, day, G, CS%interior_tendency_diags)
+
+  ! Set up memory for saved state
+  call setup_saved_state(marbl_instances%surface_flux_saved_state, G, CS%surface_flux_saved_state)
+!  call setup_saved_state(marbl_instances%interior_tendency_saved_state, G, CS%interior_tendency_saved_state)
+  ! TODO: if restarting, get value from restart instead of setting to 0!
+  do m=1, size(CS%surface_flux_saved_state)
+    if (allocated(CS%surface_flux_saved_state(m)%field_2d)) then
+      CS%surface_flux_saved_state(m)%field_2d = 0
+    else
+      CS%surface_flux_saved_state(m)%field_3d = 0
+    end if
+  end do
+
+  ! Establish location of source
+  do m= 1, CS%ntr
+    write(name(:),'(A)') trim(marbl_instances%tracer_metadata(m)%short_name)
+    OK = tracer_Z_init(CS%tr(:,:,:,m), h, CS%IC_file, name, G, US, -1e34)
+    if (.not.OK) call MOM_error(FATAL,"initialize_MARBL_tracers: "//&
+                               "Unable to read "//trim(name)//" from "//&
+                               trim(CS%IC_file)//".")
+  end do
+
+end subroutine initialize_MARBL_tracers
+
+subroutine register_MARBL_diags(MARBL_diags, diag, day, G, id_diags)
+
+  type(marbl_diagnostics_type), intent(in)    :: MARBL_diags !< MARBL diagnostics from marbl_instances
+  type(time_type), target,      intent(in)    :: day  !< Time of the start of the run.
+  type(diag_ctrl), target,      intent(in)    :: diag !< Structure used to regulate diagnostic output.
+  !integer, allocatable,         intent(inout) :: id_diags(:) !< allocatable array storing diagnostic index number
+  type(ocean_grid_type),              intent(in) :: G    !< The ocean's grid structure
+  type(temp_marbl_diag), allocatable, intent(inout) :: id_diags(:) !< allocatable array storing diagnostic index number and buffer space for collecting diags from all columns
+
+  integer :: m, diag_size
+
+  diag_size = size(MARBL_diags%diags)
+  allocate(id_diags(diag_size))
+  do m = 1, diag_size
+    id_diags(m)%id = -1
+    if (trim(MARBL_diags%diags(m)%vertical_grid) .eq. "none") then ! 2D field
+      id_diags(m)%id = register_diag_field("ocean_model", &
+        trim(MARBL_diags%diags(m)%short_name), &
+        diag%axesT1, & ! T => tracer grid? 1 => no vertical grid
+        day, &
+        trim(MARBL_diags%diags(m)%long_name), &
+        trim(MARBL_diags%diags(m)%units))
+      allocate(id_diags(m)%field_2d(SZI_(G),SZJ_(G)))
+      id_diags(m)%field_2d = 0.0
```

Comment:
Style guide says to initialize arrays with the appropriate number of parentheses `id_diags(m)%field_2d(:,:) = 0.`